Add Characteristics pages (#2054)
jpmorganchase/salt-ds: site/docs/foundations/characteristics/index.mdx

### Why characteristics?

From a design perspective, there is a need for aesthetic flexibility without decreasing the efficiency of build or the effectiveness of the components when implemented. From a dev perspective, it provides an easy way to reuse the same values, without the overhead of having to remember details such as hex values. Any time you want to tweak the value of a characteristic, you only have to do it in one place for it to reflect everywhere.

Characteristics offer the following benefits:

- Ability to theme the Salt inline with third-parties or CIB sub brand guidelines.
- Increase consistency across all components, with the knock-on effect of improving the overall user experience e.g. reduced cognitive load.
- Simplified design of new components and patterns e.g. drag-and-drop behavior.
- Streamlined evolution – a single change to a characteristic will automatically cascade down through all related components within the library.

### Token name convention

> When defining characteristic design tokens, they should generally be in the format `--salt-<characteristic name>-<variant>-<scale>-<density>`

Reviewer: ```suggestion
When defining characteristic design tokens, they should generally be in the format `--salt-<characteristic name>-<variant>-<attribute>-<state>`
```

1. Always begin with `--salt-`
2. Follow this by the foundation name, e.g. `---salt-actionable-`, `---salt-text-`, `--salt-focused-`
3. Append the variant if applicable, e.g. `--salt-actionable-cta-`, `--salt-text-h2-`
4. Append the attribute, e.g. `--salt-actionable-cta-background`, `--salt-text-h2-fontSize`, `--salt-focused-outlineColor`
5. Append the state if applicable, e.g. `--salt-actionable-cta-background-hover`

### Styling with Characteristics

Remember to wrap the characteristic you wish to use inside a `var()` function.

For example, a Button is a component on which a user can perform an action, so you could use the following characteristics to style the component:

```css
.saltButton {
  background: var(--salt-actionable-primary-background);
  font-weight: var(--salt-actionable-fontWeight);
}
```

> Characteristics are both light/dark and density scoped, meaning that the value the characteristic points to will change according to the values given in the nearest `SaltProvider`. Considering the above, we would get that in light

Reviewer: SaltProvider is in code format `SaltProvider` but there are other components mentioned below e.g. Panel, Card in plain text. how do we want to present them?

Reviewer: i'm not sure, would be good for content to take a look - cc @jpmorganchase/salt-content

theme, `--salt-actionable-primary-background` resolves to `var(--salt-palette-interact-primary-background)`, which in light theme is `var(--salt-color-gray-60)` and in dark theme is `var(--salt-color-gray-300)`, of which foundation tokens then in turn resolve to their respective RGBA values.

> ### Simple customization using characteristics

Reviewer: I think we should actually delete this whole section until we agree as a team on how this is going to be dealt with (hopefully ASAP, and this will encourage the conversation)

e.g., --saltCard-borderRadius token used doesn't even exist anymore, since this doc was written a while ago, needs to be updated. we need designers to also give input on which attributes we want to expose, if we do go down the CSS API route
cc @bhoppers2008 @pseys

Author: yea for now i had just moved over the docs from storybook,, but lmk what all needs to be removed or changed

Reviewer: just the bit from this heading to the next should be deleted, the rest is okay :) I made minor tweaks where it needed refreshing

Author: the stuff under "Simple customization using characteristics" ?

Characteristics offer the following benefits:

- Ability to theme the Salt inline with third-parties or CIB sub brand guidelines.
- Increase consistency across all components, with the knock-on effect of improving the overall user experience e.g. reduced cognitive load.
- Simplified design of new components and patterns e.g. drag-and-drop behavior.
- Streamlined evolution – a single change to a characteristic will automatically cascade down through all related components within the library.

### Token name convention

When defining characteristic design tokens, they should generally be in the format `--salt-<characteristic name>-<variant>-<scale>-<density>`

1. Always begin with `--salt-`
2. Follow this by the foundation name, e.g. `---salt-actionable-`, `---salt-text-`, `--salt-focused-`
3. Append the variant if applicable, e.g. `--salt-actionable-cta-`, `--salt-text-h2-`
4. Append the attribute, e.g. `--salt-actionable-cta-background`, `--salt-text-h2-fontSize`, `--salt-focused-outlineColor`

> 5. Append the state if applicable, e.g. `--salt-actionable-cta-background-hover`, `--salt-focused-outlineColor-warning`

Reviewer: ```suggestion
5. Append the state if applicable, e.g. `--salt-actionable-cta-background-hover`
```

### Styling with Characteristics

Remember to wrap the characteristic you wish to use inside a `var()` function.

For example, a Button is a component on which a user can perform an action, so you could use the following characteristics to style the component:

```css
.saltButton {
  background: var(--salt-actionable-primary-background);
  font-weight: var(--salt-actionable-fontWeight);
}
```

Characteristics are both light/dark and density scoped, meaning that the value the characteristic points to will change according to the values given in the nearest `SaltProvider`. Considering the above, we would get that in light

> theme, --salt-actionable-primary-background resolves to `var(--salt-action-primary-background)`, which in light theme is `var(--salt-color-gray-60)` and in dark theme is `var(--salt-color-gray-300)`, of which foundation tokens then in turn resolve to their respective RGBA values.

Reviewer: ```suggestion
theme, `--salt-actionable-primary-background` resolves to `var(--salt-action-primary-background)`, which in light theme is `var(--salt-color-gray-60)` and in dark theme is `var(--salt-color-gray-300)`, of which foundation tokens then in turn resolve to their respective RGBA values.
```

> theme, `--salt-actionable-primary-background` resolves to `var(--salt-action-primary-background)`, which in light theme is `var(--salt-color-gray-60)` and in dark theme is `var(--salt-color-gray-300)`, of which foundation tokens then in turn resolve to their respective RGBA values.

Reviewer: ```suggestion
theme, `--salt-actionable-primary-background` resolves to `var(--salt-palette-interact-primary-background)`, which in light theme is `var(--salt-color-gray-60)` and in dark theme is `var(--salt-color-gray-300)`, of which foundation tokens then in turn resolve to their respective RGBA values.
```

### Simple customization using characteristics

> In the following example, we will show how to customize the Card component through characteristics. Some components, like Button and Panel, have different variants such as 'cta' and 'secondary'. Currently, Card only has one variant, and uses tokens such as `--salt-container-primary-background` by default for all of its styling purposes. Here's how the Salt Card looks contained within a secondary Panel:

Reviewer: "Here's how the Salt Card looks contained within a secondary Panel:"

needs image following

```css
.saltCard.saltCard-secondary {
  --saltCard-background: var(--salt-container-secondary-background);
  --saltCard-borderColor: var(--salt-container-secondary-borderColor);
  --saltCard-borderRadius: 10px;
  --saltCard-borderWidth: 3px;
}
```

Note the use of `secondary` container characteristic tokens provided by Salt that weren't used in Card by default, as well as raw values to further emphasize the component. Now, all we have to do to use this secondary variant styling is pass in the `saltCard-secondary` class name we defined to Card (this name has been chosen to follow convention, but could really be anything you like):

Alternatively, we could actually redefine our design to mean that all Card components in our app convey an important message which could be accented. In this case, we'll override with accent characteristics:

```css
.my-app .saltCard {
  --saltCard-background: var(--salt-accent-background);
  --saltCard-borderColor: var(--salt-accent-borderColor);
  --saltCard-color: var(--salt-accent-foreground);
}
```

> Note how no class name is needed to be passed in. We have just redefined the Card's styling at a higher specificity, to produce the following styling:

Reviewer: needs image following

### How do I know which characteristics to apply?

Generally, the characteristics appropriate for the component should be given by the design team. To decide which characteristics to use when styling a new component, it may be useful to ask questions like:

- What can the user do with this component? Can they click it, drag it, etc?
- What state can this component be in? Can it be focused, disabled, etc?
- What is the purpose of this component? Will it contain other elements, lead me to a new page, etc?
- Should this component give feedback? Can it be submitted, does it have a success state, etc?

> It is important to note that just because you have assigned a characteristic to a component, does not mean it consumes _all_ of those characteristics. We can freely pick which characteristics we want. A good example of this may be to notice that the `focused` characteristic offers both a border and outline style. From this you would probably want to use an outline if you're setting the outline attribute, and of course border for the border attribute. Yet, there is nothing stopping you from writing `border-style: var(--salt-focused-outlineStyle)` instead.
> 
> Think carefully when deciding which characteristics to use. It's up to you to choose which characteristic is most appropriate - for example, should the background color of your component be coming from container, because that's its general purpose, or should it be coming from selectable, since it can be selected? Many characteristics offer the same attribute (like background color), and it's up to you to decide what part of your component you are styling as to which characteristic it should come from. Taking time to consider this will ensure you have the most consistent styling possible across your app.

Reviewer: ```suggestion
It is important to note that just because you have assigned a characteristic to a component, does not mean it consumes _all_ of those characteristics. We can freely pick which characteristics we want. A good example of this may be to notice that the `focused` characteristic offers both a border and outline style. From this you would want to use the outline token if you're setting the outline attribute, and the border token if setting the border attribute.

Think carefully when deciding which characteristics are most appropriate. For example, should the background color of your component be coming from container, because that's its general purpose, or should it be coming from selectable, since it can be selected? Many characteristics host for the same attribute (e.g. background color), and it's important to make the correct decision as to which characteristic your component should adopt the token from. Taking time to consider this will ensure you have the most consistent styling possible across your app, especially when different themes are applied.
```

## How do I know which characteristics the Salt components are using?

> There are two easy ways to check this, depending on what you're interested in:
> 
> - Each component has a breakdown of the characteristics used on its component story.
> - Each characteristic has a list of components using one or more of that characteristics token, which can be found at the bottom of the characteristics docs page.
> 
> You can of course also dive into the components CSS files to see exactly how they are styled.

Reviewer: ```suggestion
Each component has a breakdown of the characteristics used on its documentation page within Storybook. You can also dive into the components CSS files to see exactly how they are styled.
```

<Panel variant="secondary">
  <Card className="saltCard-secondary">
    <H1 style={{ margin: "0", lineHeight: "1.3em" }}>
      Card with our secondary styling
    </H1>
    <Text>Here is some content</Text>
  </Card>
</Panel>

Alternatively, we could actually redefine our design to mean that all Card components in our app convey an important message which could be accented. In this case, we'll override with accent characteristics:

```css
.my-app .saltCard {
  --saltCard-background: var(--salt-accent-background);
  --saltCard-borderColor: var(--salt-accent-borderColor);
  --saltCard-color: var(--salt-accent-foreground);
}
```

Note how no class name is needed to be passed in. We have just redefined the Card's styling at a higher specificity, to produce the following styling:

<Panel className="my-app" variant="secondary">
  <Card>
    <H1 style={{ margin: "0", lineHeight: "1.3em" }}>
      Card with accent styles
    </H1>
    <Text>Here is some content</Text>
  </Card>
</Panel>

> ### How do I know which characteristics to apply?

Reviewer: still need this heading
